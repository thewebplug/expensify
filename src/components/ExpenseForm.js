import React from 'react';
import moment from 'moment';
import { SingleDatePicker } from 'react-dates';



const now = moment();
console.log(moment());
class ExpenseForm extends React.Component {
    state = {
        description: this.props.currentExpense ? this.props.currentExpense.description : '',
        note: this.props.currentExpense ? this.props.currentExpense.note : '',
        amount: this.props.currentExpense ? (this.props.currentExpense.amount / 100).toString() : '',
        createdAt:  this.props.currentExpense ? moment(this.props.currentExpense.createdAt) : moment(),
        calenderFocused: false,
        error: undefined
    }

    onDescriptionChange = (e) => {
        this.setState(() => ({description: e.target.value}))
    }

    onNoteChange = (e) => {
        this.setState(() => ({note: e.target.value}))
    }
    onAmountChange = (e) => {
        const amount = e.target.value;
        if(amount.match(/^\d+(\.\d{0,2})?$/)) {
            this.setState(() => ({amount}))
        }
    }

    onDateChange = (createdAt) => {
        if(createdAt) {
            this.setState(() => ({createdAt}));
        }
    }

    onFocusChange = ({focused}) => {
        this.setState(() => ({calenderFocused: focused}));
    }

    onSubmit = (e) => {
        e.preventDefault();

        if(!this.state.description || !this.state.amount){
            const error = 'Oops! Something went wrong'
            this.setState(() =>({error}))
        }
        else{
            this.setState(() => ({error: ''}))
            this.props.onSubmit(({
                description: this.state.description,
                amount: parseFloat(this.state.amount, 10) * 100,
                createdAt: this.state.createdAt.valueOf(),
                note: this.state.note
            }))
        }
    } 


    render() {
        return(
            <div>
               {this.state.error && <p>{this.state.error}</p>}
               <form onSubmit={this.onSubmit}>
                <input 
                    type="text" 
                    placeholder="Description" 
                    autoFocus  
                    onChange={this.onDescriptionChange}
                    value={this.state.description}
                />
                <input 
                    type="number" 
                    placeholder="Amount" 
                    onChange={this.onAmountChange}
                    value={this.state.amount}
                />
                <SingleDatePicker
                    date={this.state.createdAt}
                    onDateChange={this.onDateChange}
                    focused={this.state.calenderFocused}
                    onFocusChange ={this.onFocusChange}    
                    numberOfMonths={1}
                    isOutsideRange={(day) => false}
                />
                <textarea
                    placeholder="Add a for your expense(optional)" 
                    onChange={this.onNoteChange}
                    value={this.state.note}
                    >
                </textarea>
                <button>Add Expense</button>
               </form>
            </div>
        )
        
    }
}

export default ExpenseForm;