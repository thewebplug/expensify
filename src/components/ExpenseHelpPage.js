import React from 'react';


const ExpenseHelpPage = () => (
    <div>
        Expense Help Page
    </div> 
);

export default ExpenseHelpPage;