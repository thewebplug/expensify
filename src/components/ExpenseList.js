import React from 'react';
import ExpenseListItem from './ExpenseListItem';
const ExpenseList = () => (
    <div>
      <ExpenseListItem />
    </div>
)

export default ExpenseList;
