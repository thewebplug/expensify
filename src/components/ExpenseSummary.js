import React from 'react';
import getVisibleExpenses from '../selectors/expenses';
import selectExpenseTotal from '../selectors/expense-total';
import { connect } from 'react-redux';
import numeral from 'numeral';


export const ExpenseSummary = (props) => {
    const totalAmount = selectExpenseTotal(props.expenses)
    const expenseWord = props.expenses.length === 1 ? 'expense' : 'expenses';
    return (
        <div>
            <h3>Viewing {props.expenses.length} {expenseWord} totalling {numeral(totalAmount / 100).format('$0,0.00')}</h3>
        </div>
    )
}


const mapStateToProps = (state) => {
    return{
        expenses: getVisibleExpenses(state.expenses, state.filters)
    }
}


export default connect(mapStateToProps)(ExpenseSummary);