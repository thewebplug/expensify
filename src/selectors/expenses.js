import moment from "moment";

const getVisibleExpenses = (expenses, {text, sortBy, startDate, endDate}) => {
    return expenses.filter((expense) => {
        const momentCreatedAt = moment(expense.createdAt);

        const startDateMatch = startDate ? momentCreatedAt.isSameOrAfter(startDate, 'day') : true
        const endDateMatch = endDate ? momentCreatedAt.isSameOrBefore(endDate, 'day') : true
        const textMatch = expense.description.toLowerCase().includes(text.toLowerCase());
        
        return startDateMatch && endDateMatch && textMatch;
        
    }).sort((a, b) => {
        if(sortBy === 'date'){
            return a.createdAt < b.createdAt ? 1 : -1;
        }
        else if(sortBy === 'amount') {
            return a.amount > b.amount ? -1 : 1;
        }
    })
}

export default getVisibleExpenses;