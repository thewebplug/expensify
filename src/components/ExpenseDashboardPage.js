import React from 'react';
import ExpenseList from './ExpenseList';
import ExpenseListFilters from './ExpenseListFilters';
import ExpenseSummary from './ExpenseSummary';


const ExpenseDashboardPage = () => (
    <div>
        <ExpenseList />
        <ExpenseSummary />
        <ExpenseListFilters />
    </div>
);

export default ExpenseDashboardPage;