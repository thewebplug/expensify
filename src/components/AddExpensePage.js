import React from 'react';
import ExpenseForm from './ExpenseForm';
import { connect } from 'react-redux';
import { addExpense } from '../actions/expenses';
import { useNavigate } from 'react-router-dom';

export const AddExpensePage = (props) => {
    const onSubmit=(expense) => {
        props.onSubmit(expense);
        props.navigate('/');
    }
    return(
        <div>
            <h1>Add Expenses</h1>
            <ExpenseForm onSubmit={onSubmit}/>
        </div>
    )
};

const mapDispatchToProps = (dispatch) => ({
        onSubmit: (expense) => dispatch(addExpense(expense)),
        navigate: useNavigate()
    })

export default connect(undefined, mapDispatchToProps)(AddExpensePage);