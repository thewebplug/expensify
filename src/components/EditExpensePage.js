import React from 'react';
import { useParams } from 'react-router-dom';
import { connect } from 'react-redux';
import ExpenseForm from './ExpenseForm';
import { editExpense, removeExpense } from '../actions/expenses';
import { useNavigate } from 'react-router';




export const EditExpensePage = (props) => {
        const {id} = props.useParams;

        const selectedExpense = props.expenses.find((expense) => expense.id === id);

        const onSubmit = (expense) => {
                props.editExpense(selectedExpense.id, expense)
                props.navigate('/')
        }

        const onClick = () => {
                props.removeExpense(selectedExpense.id);
                props.navigate('/')
        }
        return(
                <div>
                        <ExpenseForm
                                currentExpense={selectedExpense} 
                                onSubmit={onSubmit}
                        />
                        <button onClick={onClick}>Remove</button>
                </div>
        )
}
const mapDispatchToProps = (dispatch) => ({
        
        editExpense: (id, expense) => dispatch(editExpense(id, expense)),
        removeExpense: (id) => dispatch(removeExpense({id})),
        navigate: useNavigate(),
        useParams: useParams()
})

const mapStateToProps = (state) => ({
        expenses: state.expenses
})

export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage);