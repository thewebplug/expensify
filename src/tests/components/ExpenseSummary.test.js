import React from 'react';
import { shallow } from "enzyme";
import {ExpenseSummary} from '../../components/ExpenseSummary';
import expenses from '../fixtures/expenses'

test('should return ExpenseSummary with expenses', () => {
    const wrapper = shallow(<ExpenseSummary expenses={expenses} />)
    expect(wrapper).toMatchSnapshot();
})

test('should return ExpenseSummary without expenses', () => {
    const emptyExpense = []
    const wrapper = shallow(<ExpenseSummary expenses={emptyExpense} />)
    expect(wrapper).toMatchSnapshot();
})