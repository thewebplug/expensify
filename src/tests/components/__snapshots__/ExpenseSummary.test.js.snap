// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should return ExpenseSummary with expenses 1`] = `
<div>
  Viewing 
  3
   
  expenses
   totalling 
  $44.00
</div>
`;

exports[`should return ExpenseSummary without expenses 1`] = `
<div>
  Viewing 
  0
   
  expenses
   totalling 
  $0.00
</div>
`;
